File: apps/desktop/src/api/ollama.ts

import { OLLAMA_TIMEOUT_MS } from '../constants/timeouts';
import { invoke } from '../lib/tauri-mock';

export interface OllamaModelDetails {
  parameter_size: string;
  quantization_level: string;
  family: string;
  families: string[];
  parent_model: string;
  format: string;
}

export interface OllamaModel {
  name: string;
  size: number;
  modified_at: string;
  digest: string;
  details: OllamaModelDetails;
}

const OLLAMA_MODEL_PULL_TIMEOUT_MS = 24 * 60 * 60 * 1000;

async function invokeWithTimeout<T>(
  command: string,
  args?: Record<string, unknown>,
  timeoutMs: number = OLLAMA_TIMEOUT_MS,
): Promise<T> {
  return new Promise((resolve, reject) => {
    const timeoutId = setTimeout(() => {
      reject(new Error(`Ollama command '${command}' timed out after ${timeoutMs}ms`));
    }, timeoutMs);

    invoke<T>(command, args)
      .then((result) => {
        clearTimeout(timeoutId);
        resolve(result);
      })
      .catch((error) => {
        clearTimeout(timeoutId);
        reject(error);
      });
  });
}

export function formatModelSize(bytes: number): string {
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
  if (bytes < 1024 * 1024 * 1024) return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
  return `${(bytes / (1024 * 1024 * 1024)).toFixed(2)} GB`;
}

/**
 * Check if Ollama server is running and accessible.
 *
 * @returns true if Ollama is running and responding, false otherwise
 *
 * @example
 * ```ts
 * const isRunning = await ollamaCheckStatus();
 * if (!isRunning) {
 *   console.log('Start Ollama with: ollama serve');
 * }
 * ```
 */
export async function ollamaCheckStatus(baseUrl?: string): Promise<boolean> {
  try {
    return await invokeWithTimeout<boolean>('ollama_check_status', { baseUrl });
  } catch (error) {
    console.debug('[Ollama] Status check failed:', error);
    return false;
  }
}

/**
 * Fetch the list of installed Ollama models.
 *
 * @returns Array of installed models with metadata
 * @throws Error if Ollama is not running or the request fails
 *
 * @example
 * ```ts
 * const models = await ollamaListModels();
 * models.forEach(model => {
 *   console.log(`${model.name} (${formatModelSize(model.size)})`);
 * });
 * ```
 */
export async function ollamaListModels(baseUrl?: string): Promise<OllamaModel[]> {
  try {
    return await invokeWithTimeout<OllamaModel[]>('ollama_list_models', { baseUrl });
  } catch (error) {
    throw new Error(`Failed to list Ollama models: ${error}`);
  }
}

/**
 * Get detailed information about a specific Ollama model.
 *
 * @param modelName - The provider-reported name of an installed model
 * @returns Model details including parameters and quantization
 * @throws Error if the model is not found or Ollama is not running
 *
 * @example
 * ```ts
 * const info = await ollamaGetModelInfo(selectedModel.name);
 * console.log(`Parameters: ${info.details.parameter_size}`);
 * ```
 */
export async function ollamaGetModelInfo(
  modelName: string,
  baseUrl?: string,
): Promise<OllamaModel> {
  if (!modelName || modelName.trim().length === 0) {
    throw new Error('Model name cannot be empty');
  }

  try {
    return await invokeWithTimeout<OllamaModel>('ollama_get_model_info', {
      modelName,
      baseUrl,
    });
  } catch (error) {
    throw new Error(`Failed to get model info for '${modelName}': ${error}`);
  }
}

/**
 * Pull (download) a model from Ollama.
 * Resolves after Ollama finishes the download.
 *
 * @param modelName - The model name selected by the user
 * @throws Error if the model name is invalid or the request fails
 *
 * @example
 * ```ts
 * await ollamaPullModel(modelName);
 * console.log('Model download initiated');
 * ```
 */
export async function ollamaPullModel(modelName: string, baseUrl?: string): Promise<void> {
  if (!modelName || modelName.trim().length === 0) {
    throw new Error('Model name cannot be empty');
  }

  try {
    await invokeWithTimeout<void>(
      'ollama_pull_model',
      { modelName, baseUrl },
      OLLAMA_MODEL_PULL_TIMEOUT_MS,
    );
  } catch (error) {
    throw new Error(`Failed to pull model '${modelName}': ${error}`);
  }
}

/**
 * Delete an installed Ollama model.
 *
 * @param modelName - The name of the model to delete
 * @throws Error if the model name is invalid or the deletion fails
 *
 * @example
 * ```ts
 * await ollamaDeleteModel(selectedModel.name);
 * console.log('Model deleted successfully');
 * ```
 */
export async function ollamaDeleteModel(modelName: string, baseUrl?: string): Promise<void> {
  if (!modelName || modelName.trim().length === 0) {
    throw new Error('Model name cannot be empty');
  }

  try {
    await invokeWithTimeout<void>('ollama_delete_model', { modelName, baseUrl });
  } catch (error) {
    throw new Error(`Failed to delete model '${modelName}': ${error}`);
  }
}

/**
 * OllamaClient provides a class-based interface for Ollama operations.
 *
 * @example
 * ```ts
 * const isAvailable = await OllamaClient.checkStatus();
 * if (isAvailable) {
 *   const models = await OllamaClient.listModels();
 * }
 * ```
 */
export class OllamaClient {
  static async checkStatus(baseUrl?: string): Promise<boolean> {
    return ollamaCheckStatus(baseUrl);
  }

  static async listModels(baseUrl?: string): Promise<OllamaModel[]> {
    return ollamaListModels(baseUrl);
  }

  static async getModelInfo(modelName: string, baseUrl?: string): Promise<OllamaModel> {
    return ollamaGetModelInfo(modelName, baseUrl);
  }

  static async pullModel(modelName: string, baseUrl?: string): Promise<void> {
    return ollamaPullModel(modelName, baseUrl);
  }

  static async deleteModel(modelName: string, baseUrl?: string): Promise<void> {
    return ollamaDeleteModel(modelName, baseUrl);
  }

  static async isReadyForUse(baseUrl?: string): Promise<{
    available: boolean;
    modelCount: number;
    error?: string;
  }> {
    try {
      const isRunning = await ollamaCheckStatus(baseUrl);
      if (!isRunning) {
        return {
          available: false,
          modelCount: 0,
          error: 'Ollama is not running. Start it with: ollama serve',
        };
      }

      const models = await ollamaListModels(baseUrl);
      return {
        available: true,
        modelCount: models.length,
        error:
          models.length === 0
            ? 'No models installed. Pull one with: ollama pull <model-name>'
            : undefined,
      };
    } catch (error) {
      return {
        available: false,
        modelCount: 0,
        error: String(error),
      };
    }
  }
}

export default OllamaClient;

File: apps/desktop/src-tauri/src/sys/commands/ollama.rs
//! Ollama-specific Tauri commands for dynamic model fetching and status checking.
//!
//! This module provides commands to:
//! - Check if Ollama is running locally
//! - Fetch the list of installed Ollama models
//! - Get detailed model information

use serde::{Deserialize, Serialize};
use std::time::Duration;

/// Represents an Ollama model with its metadata
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OllamaModel {
    /// The model name (e.g., "llama3.2:latest")
    pub name: String,
    /// Size of the model in bytes
    pub size: u64,
    /// ISO 8601 timestamp of when the model was last modified
    pub modified_at: String,
    /// Model digest/hash
    #[serde(default)]
    pub digest: String,
    /// Additional model details
    #[serde(default)]
    pub details: OllamaModelDetails,
}

/// Detailed information about an Ollama model
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct OllamaModelDetails {
    /// Parameter size (e.g., "7B", "13B")
    #[serde(default)]
    pub parameter_size: String,
    /// Quantization level (e.g., "Q4_0", "Q8_0")
    #[serde(default)]
    pub quantization_level: String,
    /// Model family (e.g., "llama", "mistral")
    #[serde(default)]
    pub family: String,
    /// Model families this model belongs to
    #[serde(default)]
    pub families: Vec<String>,
    /// Parent model name
    #[serde(default)]
    pub parent_model: String,
    /// Model format
    #[serde(default)]
    pub format: String,
}

/// Response from Ollama's /api/tags endpoint
#[derive(Debug, Clone, Serialize, Deserialize)]
struct OllamaTagsResponse {
    models: Vec<OllamaApiModel>,
}

/// Internal representation of a model from Ollama's API
#[derive(Debug, Clone, Serialize, Deserialize)]
struct OllamaApiModel {
    name: String,
    #[serde(default)]
    size: u64,
    #[serde(default)]
    modified_at: String,
    #[serde(default)]
    digest: String,
    #[serde(default)]
    details: Option<OllamaApiModelDetails>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
struct OllamaApiModelDetails {
    #[serde(default)]
    parameter_size: Option<String>,
    #[serde(default)]
    quantization_level: Option<String>,
    #[serde(default)]
    family: Option<String>,
    #[serde(default)]
    families: Option<Vec<String>>,
    #[serde(default)]
    parent_model: Option<String>,
    #[serde(default)]
    format: Option<String>,
}

use crate::core::llm::OLLAMA_DEFAULT_BASE_URL;

fn resolve_ollama_base_url(base_url: Option<String>) -> Result<String, String> {
    let candidate = base_url
        .as_deref()
        .unwrap_or(OLLAMA_DEFAULT_BASE_URL)
        .trim()
        .trim_end_matches('/');
    let parsed =
        url::Url::parse(candidate).map_err(|_| "Ollama URL must be a valid URL".to_string())?;

    if !matches!(parsed.scheme(), "http" | "https") {
        return Err("Ollama URL must use http or https".to_string());
    }
    if parsed.host_str().is_none() {
        return Err("Ollama URL must include a host".to_string());
    }
    if !parsed.username().is_empty() || parsed.password().is_some() {
        return Err("Ollama URL must not contain credentials".to_string());
    }
    if parsed.query().is_some() || parsed.fragment().is_some() {
        return Err("Ollama URL must not contain a query or fragment".to_string());
    }

    Ok(candidate.to_string())
}

/// Check if Ollama server is running and accessible
///
/// # Returns
/// - `Ok(true)` if Ollama is running and responding
/// - `Ok(false)` if Ollama is not running or not accessible
#[tauri::command]
pub async fn ollama_check_status(base_url: Option<String>) -> Result<bool, String> {
    let base_url = resolve_ollama_base_url(base_url)?;
    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(5))
        .build()
        .map_err(|e| format!("Failed to create HTTP client: {}", e))?;

    match client
        .get(format!("{}/api/tags", base_url))
        .timeout(Duration::from_secs(3))
        .send()
        .await
    {
        Ok(response) => Ok(response.status().is_success()),
        Err(_) => Ok(false),
    }
}

/// Fetch the list of installed Ollama models
///
/// # Returns
/// - `Ok(Vec<OllamaModel>)` with the list of installed models
/// - `Err` if Ollama is not running or the request fails
#[tauri::command]
pub async fn ollama_list_models(base_url: Option<String>) -> Result<Vec<OllamaModel>, String> {
    let base_url = resolve_ollama_base_url(base_url)?;
    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(15))
        .build()
        .map_err(|e| format!("Failed to create HTTP client: {}", e))?;

    let response = client
        .get(format!("{}/api/tags", base_url))
        .timeout(Duration::from_secs(10))
        .send()
        .await
        .map_err(|e| {
            if e.is_connect() {
                "Ollama is not running. Start it with 'ollama serve' in your terminal.".to_string()
            } else if e.is_timeout() {
                "Connection to Ollama timed out. Please check if Ollama is running.".to_string()
            } else {
                format!("Failed to connect to Ollama: {}", e)
            }
        })?;

    if !response.status().is_success() {
        return Err(format!(
            "Ollama API returned error status: {}",
            response.status()
        ));
    }

    let tags_response: OllamaTagsResponse = response
        .json()
        .await
        .map_err(|e| format!("Failed to parse Ollama response: {}", e))?;

    let models = tags_response
        .models
        .into_iter()
        .map(|m| {
            let details = m.details.unwrap_or_default();
            OllamaModel {
                name: m.name,
                size: m.size,
                modified_at: m.modified_at,
                digest: m.digest,
                details: OllamaModelDetails {
                    parameter_size: details.parameter_size.unwrap_or_default(),
                    quantization_level: details.quantization_level.unwrap_or_default(),
                    family: details.family.unwrap_or_default(),
                    families: details.families.unwrap_or_default(),
                    parent_model: details.parent_model.unwrap_or_default(),
                    format: details.format.unwrap_or_default(),
                },
            }
        })
        .collect();

    Ok(models)
}

/// Get detailed information about a specific Ollama model
///
/// # Arguments
/// * `model_name` - The name of the model to get details for (e.g., "llama3.2:latest")
///
/// # Returns
/// - `Ok(OllamaModel)` with the model details
/// - `Err` if the model is not found or Ollama is not running
#[tauri::command]
pub async fn ollama_get_model_info(
    model_name: String,
    base_url: Option<String>,
) -> Result<OllamaModel, String> {
    let models = ollama_list_models(base_url).await?;

    models
        .into_iter()
        .find(|m| m.name == model_name || m.name.starts_with(&format!("{}:", model_name)))
        .ok_or_else(|| format!("Model '{}' not found in Ollama", model_name))
}

/// Pull a model from Ollama and wait for the download to complete.
///
/// # Arguments
/// * `model_name` - The name of the model to pull (e.g., "llama3.2", "mistral:7b")
///
/// # Returns
/// - `Ok(())` when the model is ready
/// - `Err` if the request fails
#[tauri::command]
pub async fn ollama_pull_model(model_name: String, base_url: Option<String>) -> Result<(), String> {
    if model_name.trim().is_empty() {
        return Err("Model name cannot be empty".to_string());
    }
    let base_url = resolve_ollama_base_url(base_url)?;

    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(86400))
        .build()
        .map_err(|e| format!("Failed to create HTTP client: {}", e))?;

    let response = client
        .post(format!("{}/api/pull", base_url))
        .json(&serde_json::json!({
            "name": model_name,
            "stream": false
        }))
        .timeout(Duration::from_secs(86400))
        .send()
        .await
        .map_err(|e| format!("Failed to initiate model pull: {}", e))?;

    if !response.status().is_success() {
        let error_text = response.text().await.unwrap_or_default();
        return Err(format!("Failed to pull model: {}", error_text));
    }

    Ok(())
}

/// Delete an Ollama model
///
/// # Arguments
/// * `model_name` - The name of the model to delete
///
/// # Returns
/// - `Ok(())` if the model was deleted
/// - `Err` if the deletion fails
#[tauri::command]
pub async fn ollama_delete_model(
    model_name: String,
    base_url: Option<String>,
) -> Result<(), String> {
    if model_name.trim().is_empty() {
        return Err("Model name cannot be empty".to_string());
    }
    let base_url = resolve_ollama_base_url(base_url)?;

    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(60))
        .build()
        .map_err(|e| format!("Failed to create HTTP client: {}", e))?;

    let response = client
        .delete(format!("{}/api/delete", base_url))
        .json(&serde_json::json!({
            "name": model_name
        }))
        .timeout(Duration::from_secs(30))
        .send()
        .await
        .map_err(|e| format!("Failed to delete model: {}", e))?;

    if !response.status().is_success() {
        let error_text = response.text().await.unwrap_or_default();
        return Err(format!("Failed to delete model: {}", error_text));
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_ollama_check_status_when_not_running() {
        // A running local Ollama returns true; an absent one returns false.
        // Either way the command must not turn connection refusal into an error.
        let result = ollama_check_status(None).await;
        assert!(result.is_ok());
    }

    #[test]
    fn resolves_configured_ollama_url_without_trailing_slash() {
        let url = resolve_ollama_base_url(Some("http://192.168.1.20:11434/".to_string())).unwrap();
        assert_eq!(url, "http://192.168.1.20:11434");
    }

    #[test]
    fn rejects_ollama_urls_with_embedded_credentials() {
        let result =
            resolve_ollama_base_url(Some("http://user:secret@localhost:11434".to_string()));
        assert_eq!(
            result.unwrap_err(),
            "Ollama URL must not contain credentials"
        );
    }

    #[test]
    fn test_ollama_model_serialization() {
        let model = OllamaModel {
            name: "llama3.2:latest".to_string(),
            size: 4_000_000_000,
            modified_at: "2024-01-15T10:30:00Z".to_string(),
            digest: "abc123".to_string(),
            details: OllamaModelDetails {
                parameter_size: "7B".to_string(),
                quantization_level: "Q4_0".to_string(),
                family: "llama".to_string(),
                families: vec!["llama".to_string()],
                parent_model: String::new(),
                format: "gguf".to_string(),
            },
        };

        let json = serde_json::to_string(&model).unwrap();
        assert!(json.contains("llama3.2:latest"));
        assert!(json.contains("7B"));
    }
}
